resources/resources.go

```diff
@@ -89,23 +121,45 @@ func (res *EC2NomadResource) Scale(desiredNomadCount int, vc *VaultClient) error
 
 	// scale accordingly
 	if count < desiredNomadCount {
+		// If scaling out before the timer to reset an allocation, cancel the timer.
+		if res.nextResetTime.Before(maxTime) {
+			res.nextResetTime = maxTime
+			logging.Info("[restart log] stopping timer to cancel allocation")
+		}
 		return res.scaleOut(desiredNomadCount, vc)
 	} else if count > desiredNomadCount {
+		// When doing the last scaleIn to the minimum count, start the timer to reset any allocation.
+		// This is primarily targeted towards Nomad jobs running Spark workers.
+		if desiredNomadCount == res.NomadClient.MinCount {
+			resetDuration := res.getResetDuration()
+			res.nextResetTime = now.Add(resetDuration)
+			logging.Info("[restart log] starting timer to cancel allocation in %v", resetDuration)
+		}
 		return res.scaleIn(desiredNomadCount, vc)
 	} else {
 		return nil
 	}
 }
 
-// scaleOut - scale out - ec2 then nomad
-func (res *EC2NomadResource) scaleOut(desiredNomadCount int, vc *VaultClient) error {
-	now := time.Now()
-	if time.Since(res.lastScaledTime) < res.ScaleOutCooldown {
-		return fmt.Errorf("Too soon to scale again")
+// Returns the duration of 2 cycles of ScaleOutCoolDown or 5 minutes
+// (to give a buffer for ASG to terminate EC2 instance completely), whichever is greater
+func (res *EC2NomadResource) getResetDuration() time.Duration {
+	var duration time.Duration
+	// Wait for at least 2 cycle of scaleOut
+	if res.ScaleOutCooldown > res.ScaleInCooldown {
+		duration = 2 * res.ScaleOutCooldown
+	} else {
+		duration = 2 * res.ScaleInCooldown
+	}
+	if duration < 5*time.Minute {
+		duration = 5 * time.Minute
 	}
```

Comment:
Nit: Can use `math.Min` twice also. Should work on `time.Duration` since it's just a type alias to `int64`.

Reply:
Tried in a playground and i would have to cast it to `float64` explicitly to pass in and then cast the result back to a `time.Duration` so i think this is more readable.

Reply:
:frowning_face: You are totally right, `math.Min` is float only, lack of generics tsk tsk.
So just ignore my suggestion haha